src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexPrefixAnalyzer.cs

```diff
@@ -601,94 +590,116 @@ public static (RegexNode LoopNode, (char Char, string? String, char[]? Chars) Li
+                        return node.Kind;
+
+                    case RegexNodeKind.Atomic:
+                    case RegexNodeKind.Capture:
+                        // For groups, continue exploring the sole child.
+                        node = node.Child(0);
+                        continue;
 
                     case RegexNodeKind.Concatenate:
-                        if (curNode.ChildCount() > 0)
+                        // For concatenations, we expect primarily to explore its last child, but we can also skip over
+                        // certain kinds of nodes (e.g. Empty), and thus iterate through its children backward looking for the
+                        // last we shouldn't skip.
                         {
-                            concatNode = curNode;
-                            nextChild = 0;
+                            int childCount = node.ChildCount();
+                            RegexNode? child = null;
+                            if (leading)
+                            {
+                                for (int i = 0; i < childCount; i++)
+                                {
+                                    if (node.Child(i).Kind is not (RegexNodeKind.Empty or RegexNodeKind.PositiveLookaround or RegexNodeKind.NegativeLookaround))
+                                    {
+                                        child = node.Child(i);
+                                        break;
+                                    }
+                                }
+                            }
+                            else
+                            {
+                                for (int i = childCount - 1; i >= 0; i--)
+                                {
+                                    if (node.Child(i).Kind is not (RegexNodeKind.Empty or RegexNodeKind.PositiveLookaround or RegexNodeKind.NegativeLookaround))
+                                    {
+                                        child = node.Child(i);
+                                        break;
+                                    }
+                                }
+                            }
+
+                            if (child is not null)
+                            {
+                                node = child;
+                                continue;
+                            }
+
+                            goto default;
                         }
-                        break;
 
-                    case RegexNodeKind.Atomic:
-                    case RegexNodeKind.Capture:
-                        curNode = curNode.Child(0);
-                        concatNode = null;
-                        continue;
+                    case RegexNodeKind.Alternate:
+                        // For alternations, every branch needs to lead or trail with the same anchor.
+                        {
+                            // Get the leading/trailing anchor of the first branch.  If there isn't one, bail.
```

Comment:
In case `leading == false`, don't we also want to ensure that all children's length is the same? If not, then I don't think we can make a FindFirstChar optimization to jump to a specific place since there could be different possible starting positions.

Reply:
As I said here https://github.com/dotnet/runtime/pull/64177#discussion_r791063073 I understand how you handle this now, so you can disregard this one.